components/use-vault.js

```diff
@@ -0,0 +1,313 @@
+import { useCallback, useState, useEffect } from 'react'
+import { useMe } from '@/components/me'
+import { useMutation, useQuery } from '@apollo/client'
+import { GET_ENTRY, SET_ENTRY, UNSET_ENTRY, CLEAR_VAULT, SET_VAULT_KEY_HASH } from '@/fragments/vault'
+
+export function useVaultConfigurator () {
+  const { me } = useMe()
+  const [setVaultKeyHash] = useMutation(SET_VAULT_KEY_HASH)
+
+  // vault key stored locally
+  const [vaultKey, innerSetVaultKey] = useState(null)
+
+  useEffect(() => {
+    if (!me) return
+
+    let localVaultKey = getLocalKey(me.id)
+
+    if (!me.privates.vaultKeyHash || localVaultKey?.hash !== me.privates.vaultKeyHash) {
+      // We can tell that another device has reset the vault if the values
+      // on the server are encrypted with a different key or no key exists anymore.
+      // In that case, our local key is no longer valid and our device needs to be connected
+      // to the vault again by entering the correct passphrase.
+      localVaultKey = null
+      unsetLocalKey(me.id)
+    }
+
+    innerSetVaultKey(localVaultKey)
+  }, [me?.privates?.vaultKeyHash])
+
+  // clear vault: remove everything and reset the key
+  const [clearVault] = useMutation(CLEAR_VAULT, {
+    onCompleted: () => {
+      unsetLocalKey(me.id)
+      innerSetVaultKey(null)
+    }
+  })
+
+  // initialize the vault and set a vault key
+  const setVaultKey = useCallback(async (passphrase) => {
+    const { key, hash } = await deriveKey(me.id, passphrase)
+    await setVaultKeyHash({
+      variables: { hash },
+      onError: (error) => {
+        const errorCode = error.graphQLErrors[0]?.extensions?.code
+        if (errorCode === 'VAULT_KEY_ALREADY_SET') {
+          throw new Error('Device sync is already enabled with a different key. Please input the correct key or reset the device sync.')
+        } else {
+          throw new Error(error)
+        }
+      }
+    })
+    innerSetVaultKey({ passphrase, key, hash })
+    setLocalKey(me.id, { passphrase, key, hash })
+  }, [setVaultKeyHash])
+
+  // disconnect the user from the vault (will not clear or reset the passphrase, use clearVault for that)
+  const disconnectVault = useCallback(() => {
+    unsetLocalKey(me.id)
+    innerSetVaultKey(null)
+  }, [unsetLocalKey, innerSetVaultKey])
+
+  return [vaultKey, setVaultKey, clearVault, disconnectVault]
+}
+
+export function useVaultMigration () {
+  const { me } = useMe()
+  const [setVaultEntry] = useMutation(SET_ENTRY)
+
+  // migrate local storage to vault
+  const migrate = useCallback(async () => {
+    const vaultKey = getLocalKey(me.id)
+    if (!vaultKey) throw new Error('vault key not found')
+
+    let migratedCount = 0
+
+    for (const storageKey of Object.keys(window.localStorage)) {
+      if (!storageKey.startsWith('wallet:')) continue
+      if (storageKey.includes(':local-only:')) continue
```

Comment:
the new code doesn't check if the key belongs to the current user id

Reply:
It doesn't need to check because it can assume whoever calls `useVault` knows what they are doing